File: hackerrank/problem solving/algorithms/2 - implementation/13 - counting valleys/with morfo library/counting_valleys.py
# https://www.hackerrank.com/challenges/counting-valleys/problem?isFullScreen=true

import morfo


def main():
    morfo.skip_input_ln()
    steps = morfo.read(str)
    print(counting_valleys(steps))


# n: length of steps
# T: O(n)
# S: O(1) extra space
def counting_valleys(steps):
    valleys = 0
    current_altitude = 0

    for step in steps:
        was_below_sea_level = current_altitude < 0
        current_altitude += -1 if step == 'D' else 1
        is_in_sea_level_from_valley = was_below_sea_level and current_altitude == 0
        if is_in_sea_level_from_valley:
            valleys += 1

    return valleys


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/13 - counting valleys/with morfo library/counting_valleys.ts
// https://www.hackerrank.com/challenges/counting-valleys/problem?isFullScreen=true

import { read, skipInputLn } from 'morfo';

function main() {
  skipInputLn();
  const steps: string = read(String);
  console.log(countingValleys(steps));
}

// n: length of steps
// T: O(n)
// S: O(1) extra space
function countingValleys(steps: string): number {
  let valleys = 0;
  let currentAltitude = 0;

  for (const step of steps) {
    const wasBelowSeaLevel: boolean = currentAltitude < 0;
    currentAltitude += step === 'D' ? -1 : 1;
    const isInSeaLevelFromValley: boolean =
      wasBelowSeaLevel && currentAltitude === 0;
    if (isInSeaLevelFromValley) valleys++;
  }

  return valleys;
}

main();
